win_lineinfile backrefs bug fix and updated examples.

johnnysheppard-isode opened the pull request:
##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
win_lineinfile

##### ANSIBLE VERSION
```
ansible 2.3.0 (devel ba39d1158c) last updated 2017/02/01 11:36:04 (GMT +100)
  config file = /home/js/.ansible.cfg
  configured module search path = Default w/o overrides

```

##### SUMMARY
A small fix to win_lineinfile to help it be idempotent. Also updated the docs to give an example of using **backrefs**.

- changed $backrefs to a bool so it works with true/false/yes/no.

This is based on an issue reported in the old core modules repository: https://github.com/ansible/ansible-modules-core/issues/3381

johnnysheppard-isode committed:
Bug Fix for win_lineinfile and updated examples.

- changed $backrefs to a bool so it works with true/false/yes/no. This also fixes idempotency.
- Updated Docs with an example of using backrefs.

jhawkesworth commented on lib/ansible/modules/windows/win_lineinfile.ps1:
```diff
@@ -29,7 +29,7 @@ $path= Get-Attr $params "path" $FALSE;
 $state = Get-Attr $params "state" "present";
 $line = Get-Attr $params "line" $FALSE;
-$backrefs = Get-Attr $params "backrefs" "no";
+$backrefs = ConvertTo-Bool (Get-AnsibleParam -obj $params -name "backrefs" -default "no")
```
in latest powershell.ps1 in devel you can now use -type "bool" which returns you a boolean, so you can avoid the messy looking ConvertTo-Bool call here.

jhawkesworth commented on lib/ansible/modules/windows/win_lineinfile.ps1:
```diff
@@ -95,10 +95,10 @@ function WriteLines($outlines, $path, $linesep, $encodingobj, $validate) {
 	# Commit changes to the path
 	$cleanpath = $path.Replace("/", "\");
-	Copy-Item $temppath $cleanpath -force;	
+	Copy-Item $temppath $cleanpath -force;
```
consider adding try .. .catch block to capture any error messages and fail gracefully.

johnnysheppard-isode committed:
Made suggested updates and converted two more parameters to "bool"

johnnysheppard-isode commented:
Thanks @jhawkesworth for the suggestions. I've updated the PR with those suggestions and also made a couple of other small mods too to update some more parameters that should be "bool" types .

ready_for_review

jhawkesworth commented on lib/ansible/modules/windows/win_lineinfile.ps1:
```diff
@@ -95,11 +95,18 @@ function WriteLines($outlines, $path, $linesep, $encodingobj, $validate) {
+		Copy-Item $temppath $cleanpath -force;
+	}
+	catch {
+		Fail-Json "Cannot write to: $cleanpath"
```
I should have been clearer... but it would be great if this returned the actual exception message as well as your generic message.  After all, there's lots of possible reasons why you might not be able to copy - this gives users a clue where to go looking to fix the problem.  (sometimes - I find some microsoft error messages to be confusing).

Something like
Fail-Json "Cannot write to: $cleanpath. (" + $_.Exception.Message + ")"

jhawkesworth commented:
@johnnysheppard-isode  - nice - I think the booleans improve the readability of the code.  Just one tiny change noted above and I think this would be good to go.

johnnysheppard-isode committed:
Updated the Exception message

- Now contains the Windows Exception message as well as a custom message to help point in the right direction of a failed write.

johnnysheppard-isode commented:
Thanks once again for your help @jhawkesworth. New changes made as suggested :smiley:

jhawkesworth commented:
Thanks for fixing it up.  Just spotted another tiny thing though... is it worth also wrapping up the Remove-Item in try catch for the same reasons?  Sorry to pick nits, just want to make it as good as we can.

Jon

johnnysheppard-isode committed:
Updated Exception Handling

- Added Exception checks for Creating and removing the temporary files.
- Changed the ErrorAction on the copy tmpfile and remove tmp file to "Stop" to
  cause the exception handler to catch all errors so we can fail gracefully
  every time.

johnnysheppard-isode commented:
Changes made, I also added a check to the temp file creation, as there's always a chance there's an issue creating it in the first place.

Cheers,

Johnny 😃

dagwieers commented:
shipit

jhawkesworth commented:
bot_status

mattclay commented:
Closing and re-opening to trigger CI.

dagwieers commented:
@johnnysheppard-isode Not directly related to your changes, or required in this PR, but we do need all our (Windows) modules to be idempotent and support check-mode and diff output, including this one.

johnnysheppard-isode commented:
Hi @dagwieers,

I'd be happy to update the module in a separate PR to make it fully idempotent, support check-mode and diff output if you'd like, once this PR has gone through.

Is that ok?

dagwieers commented:
@johnnysheppard-isode Sure, all fine by me.

PS I am just a contributor like you, so I am not representing the project. But it is my opinion that we should no longer accept modules that are not idempotent or not include check-mode. It really ought to become mandatory, and we still have a long way to go. Diff support is nice to have (but often hard to do, or unclear how to do it).

jhawkesworth commented:
closing and opening to trigger CI build

jhawkesworth commented:
shipit

jhawkesworth commented:
bot_status
